bin/publish

```diff
@@ -1,24 +1,51 @@
 #!/bin/bash -e
 
-DOCKERHUB_IMAGE_OLD='cyberark/conjur-kubernetes-authenticator'  # backwards-compatible image name
-DOCKERHUB_IMAGE_NEW='cyberark/conjur-authn-k8s-client'
-VERSION_TAG="$(<VERSION)"
+. bin/build_utils
 
 REDHAT_IMAGE='scan.connect.redhat.com/ospid-1c46a2de-1d88-40e6-a433-7114ad0099cb/conjur-openshift-authenticator-client'
 
-docker tag conjur-authn-k8s-client:dev "$DOCKERHUB_IMAGE_OLD"
-docker tag conjur-authn-k8s-client:dev "$DOCKERHUB_IMAGE_OLD:$VERSION_TAG"
-
-docker tag conjur-authn-k8s-client:dev "$DOCKERHUB_IMAGE_NEW"
-docker tag conjur-authn-k8s-client:dev "$DOCKERHUB_IMAGE_NEW:$VERSION_TAG"
-
-docker push $DOCKERHUB_IMAGE_OLD
-docker push "$DOCKERHUB_IMAGE_OLD:$VERSION_TAG"
-
-docker push $DOCKERHUB_IMAGE_NEW
-docker push "$DOCKERHUB_IMAGE_NEW:$VERSION_TAG"
+readonly REGISTRY="cyberark"
+readonly TAG="$(version_tag)"
+readonly INTERNAL_REGISTRY="registry.tld"
+readonly VERSION="$(cat VERSION)"
+
+readonly IMAGES=(
+  "conjur-kubernetes-authenticator"
+  "conjur-authn-k8s-client"
+)
+
+readonly TAGS=(
+  "$VERSION"
+  "latest"
+)
+
+# fetching tags is required for git_description to work
+git fetch --tags
+git_description=$(git describe)
+
+for image_name in "${IMAGES[@]}"; do
+  # always push the tag with the commit hash to internal registry
+   echo "Tagging $INTERNAL_REGISTRY/$image_name:${TAG}"
+   docker tag "$image_name:$TAG" "$INTERNAL_REGISTRY/$image_name:$TAG"
+   echo "Pushing $INTERNAL_REGISTRY/$image_name:${TAG}"
+   docker push "$INTERNAL_REGISTRY/$image_name:$TAG"
+
+  # if it’s not a tagged commit, VERSION will have extra characters (i.e. -g666c4b2), so that commit won't be published
+  # only when tag matches the VERSION, push VERSION and latest releases
+  # and x and x.y releases
+  if [ "$git_description" = "v${VERSION}" ]; then
+    echo "Revision $git_description matches version $VERSION exactly. Pushing to Dockerhub..."
+
+    for tag in "${TAGS[@]}" $(gen_versions $VERSION); do
+      echo "Tagging and pushing $REGISTRY/$image_name:$tag"
+      docker tag "$image_name:$TAG" "$REGISTRY/$image_name:$tag"
+      docker push "$REGISTRY/$image_name:$tag"
+    done
+  fi
+done
 
 docker tag conjur-authn-k8s-client:dev-redhat "$REDHAT_IMAGE"
-docker tag conjur-authn-k8s-client:dev-redhat "$REDHAT_IMAGE:$VERSION_TAG"
+docker tag conjur-authn-k8s-client:dev-redhat "$REDHAT_IMAGE:$VERSION"
+
+docker push "$REDHAT_IMAGE:$VERSION" || { echo 'Red Hat push FAILED!'; exit 0; }  # you can't push the same tag twice to redhat registry, so ignore errors
```

Comment:
I'd like @sgnn7 's input as well, but my only nit here is the several different methods used for string interpolation. If my memory serves, we always try to use this format:
"${variable}"

Reply:
Yeah this one should be enclosed in braces since it seems like it would break things otherwise. I think also in this case, a real `if` would be easier to read/maintain:
```bash
if ! docker push "${REDHAT_IMAGE}:${VERSION}"; then
  echo 'Red Hat push FAILED!'
  exit 0
fi
```

Reply:
Just changed that